analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs

```diff
@@ -0,0 +1,133 @@
+        {
+            private const string TwoPartPatternWithVariableSlash = @"^[\/\\]{0}[\/\\]{1}([\/\\]|$)";
+            private const string TwoPartPatternWithForwardSlash = @"^\/{0}\/{1}(\/|$)";
+            private const string ThreePartPatternWithForwardSlash = @"^\/{0}\/{1}\/{2}(\/|$)";
+            private const string TmpRegex = @"^([a-zA-z]:)?[\\\/]?[\/\\]tmp([\/\\]|$)";
+            private const string TempRegex = @"^([a-zA-z]:)?[\\\/]?[\/\\]temp([\/\\]|$)";
+            private const string WindowsTempRegex = @"^([a-zA-z]:)?[\\\/]?[\\\/]windows[\\\/]temp([\\\/]|$)";
+            private const string UserProfileRegex = @"%USERPROFILE%\\AppData\\Local\\Temp([\\\/]|$)";
+            private const string EnvironmentVariableTemplate = @"%{0}%([\\\/]|$)";
+            private static readonly string VarTmpRegex = string.Format(TwoPartPatternWithVariableSlash, "var", "tmp");
+            private static readonly string UsrTmpRegex = string.Format(TwoPartPatternWithVariableSlash, "usr", "tmp");
+            private static readonly string DevShmRegex = string.Format(TwoPartPatternWithVariableSlash, "dev", "shm");
+            private static readonly string DevMqueueRegex = string.Format(TwoPartPatternWithForwardSlash, "dev", "mqueue");
+            private static readonly string RunLockRegex = string.Format(TwoPartPatternWithForwardSlash, "run", "lock");
+            private static readonly string VarRunLockRegex = string.Format(ThreePartPatternWithForwardSlash, "var", "run", "lock");
+            private static readonly string LibraryCachesRegex = string.Format(TwoPartPatternWithForwardSlash, "library", "caches");
+            private static readonly string UsersSharedRegex = string.Format(TwoPartPatternWithForwardSlash, "users", "shared");
+            private static readonly string PrivateTmpRegex = string.Format(TwoPartPatternWithForwardSlash, "private", "tmp");
+            private static readonly string PrivateVarTmpRegex = string.Format(ThreePartPatternWithForwardSlash, "private", "var", "tmp");
+            private static readonly string TempEnvVariable = string.Format(EnvironmentVariableTemplate, "temp");
+            private static readonly string TmpEnvVariable = string.Format(EnvironmentVariableTemplate, "tmp");
+            private static readonly string TmpDirEnvVariable = string.Format(EnvironmentVariableTemplate, "tmpdir");
+
+            private static readonly Regex LinuxPubliclyWritabelDirs =
+                new Regex($"{DevMqueueRegex}|{RunLockRegex}|{VarRunLockRegex}",
+                    RegexOptions.Compiled);
+
+            private static readonly Regex WindowsAndMacPubliclyWritabelDirs =
+                new Regex($"{TmpRegex}|{VarTmpRegex}|{UsrTmpRegex}|{DevShmRegex}|{LibraryCachesRegex}|{UsersSharedRegex}|{PrivateTmpRegex}|{PrivateVarTmpRegex}|{TempRegex}|{WindowsTempRegex}",
+                    RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
```

Comment:
I've asked to isolate few new regexes separately. Then we can use a constant to avoid this repetition 4 or 5 times (it's fine to have it twice).